core/src/Kokkos_ScratchSpace.hpp

```diff
@@ -91,6 +91,8 @@ class ScratchMemorySpace {
   typedef typename ExecSpace::array_layout array_layout;
   typedef typename ExecSpace::size_type size_type;
 
+  static constexpr const char* name() { return "ScratchMemorySpace"; }
```

Comment:
This is fine, but we don't need to scatter `::name()` all over the code just because `name()` happens to be static here. We can also get to it through the instance method

Reply:
Which of the specific changes proposed here do you have problem with?
They all were calling the default constructor to call `name()`.

Reply:
Yeah, and there's nothing wrong with that. We're allowing ourselves the freedom to make this an instance method in the future. The cost of default construct plus instance member is the same as calling the class member, so why not give ourselves that freedom in the future?

Reply:
The default constructor for `ScratchMemorySpace` is private so we would need to address that as well here. 
We are not arguing about `name()` being `static` or not. Why would choosing to call it as class member restrict us?

Reply:
> We're allowing ourselves the freedom to make this an instance method in the future.

Fair enough.  @masterleinad would you please drop all changes but the addition of the member function to `ScratchMemorySpace`

Reply:
> We are not arguing about name() being static or not.

True, @dhollman it is not about how WE call it, if we declare it to be `static` we can't prevent users to call it as a static function.

Reply:
Yeah good point. We probably should make it an instance member, actually, to keep people from calling it as a static function wherever possible.

Also, why is the default constructor private? Do we really want to remove `default_constructible` from the memory space guarantees? If we can't do anything reasonable for the `ScratchMemorySpace` default constructor, then we probably should stop using default constructors in generic contexts, but that's a much bigger problem. Right now, all models of `Space` have default constructors.

Reply:
> Yeah good point. We probably should make it an instance member, actually, to keep people from calling it as a static function wherever possible.

You mean public static data member instead of a member function?

Reply:
> Also, why is the default constructor private? Do we really want to remove `default_constructible` from the memory space guarantees? If we can't do anything reasonable for the `ScratchMemorySpace` default constructor, then we probably should stop using default constructors in generic contexts, but that's a much bigger problem. Right now, all models of `Space` have default constructors.

I agree but I don't think it should be fixed in that PR.

Reply:
We could initialize all the sizes to 0 and the pointers to `nullptr`s if you feel strongly about it.
In my opinion, there is no strong reason to make it instance member function now and would make all class member functions `static` that can be.

Reply:
All the `name()` functions of spaces are static member functions. This needs to be a static member function too. And no we cannot change this to a non-static member function, this is part of the public interface and is used in user code.